cpp/src/arrow/allocator-test.cc

```diff
@@ -0,0 +1,73 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gtest/gtest.h"
+#include "arrow/allocator.h"
+#include "arrow/test-util.h"
+#include <iostream>
+
+namespace arrow {
+
+TEST(stl_allocator, MemoryTracking) {
+  auto pool = default_memory_pool();
+  stl_allocator<uint64_t> alloc;
+  uint64_t* data = alloc.allocate(100);
+
+  ASSERT_EQ(100 * sizeof(uint64_t), pool->bytes_allocated());
+
+  alloc.deallocate(data, 100);
+  ASSERT_EQ(0, pool->bytes_allocated());
+}
+
+TEST(stl_allocator, TestOOM) {
```

Comment:
Looks like this is failing in Valgrind -- move this test case into the if block below?

1: [ RUN      ] stl_allocator.TestOOM
1: ==10654== Argument 'size' of function memalign has a fishy (possibly negative) value: -8
1: ==10654==    at 0x4C2D110: memalign (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
1: ==10654==    by 0x4C2D227: posix_memalign (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
1: ==10654==    by 0x446C95: arrow::(anonymous namespace)::AllocateAligned(long, unsigned char**) (memory_pool.cc:48)
1: ==10654==    by 0x446FBF: arrow::DefaultMemoryPool::Allocate(long, unsigned char**) (memory_pool.cc:78)
1: ==10654==    by 0x442245: arrow::stl_allocator<unsigned long>::allocate(unsigned long, void const*) (allocator.h:59)
1: ==10654==    by 0x441533: arrow::stl_allocator_TestOOM_Test::TestBody() (allocator-test.cc:39)
1: ==10654==    by 0x46AFF7: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2078)
1: ==10654==    by 0x466791: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2114)
1: ==10654==    by 0x44DD7A: testing::Test::Run() (gtest.cc:2151)
1: ==10654==    by 0x44E508: testing::TestInfo::Run() (gtest.cc:2326)
1: ==10654==    by 0x44EB43: testing::TestCase::Run() (gtest.cc:2444)
1: ==10654==    by 0x4552FE: testing::internal::UnitTestImpl::RunAllTests() (gtest.cc:4315)
1: ==10654== 
```